app/team.go

```diff
@@ -223,8 +227,17 @@ func (a *App) AddUserToTeamByHash(userId string, hash string, data string) (*mod
 		return nil, model.NewAppError("JoinUserToTeamByHash", "api.user.create_user.signup_link_invalid.app_error", nil, "", http.StatusBadRequest)
 	}
 
-	t, timeErr := strconv.ParseInt(props["time"], 10, 64)
-	if timeErr != nil || model.GetMillis()-t > 1000*60*60*48 { // 48 hours
+	result := <-a.Srv.Store.Token().GetByToken(props["token"])
+	if result.Err != nil {
+		return nil, model.NewAppError("JoinUserToTeamByHash", "api.user.create_user.signup_link_invalid.app_error", nil, result.Err.Error(), http.StatusBadRequest)
+	}
+
+	token := result.Data.(*model.Token)
+	if token.Type != TOKEN_TYPE_TEAM_INVITATION {
+		return nil, model.NewAppError("JoinUserToTeamByHash", "api.user.create_user.signup_link_invalid.app_error", nil, "", http.StatusBadRequest)
+	}
+
+	if model.GetMillis()-token.CreateAt >= TEAM_INVITATION_EXPIRY_TIME {
```

Comment:
Expired tokens won't be deleted. Same if there are any errors below. (defer the token deletion?)

Reply:
It makes sense to remove the expired tokens, but I though that was already done by another gorouting periodically. There are other tokens that can be expired (reset password, email verification). You can ask for a reset password, and never reset it. If we have to delete expired tokens I prefer to do that as a separated block of code in another PR.

Reply:
Ya the tokens are cleaned up automatically, but I just thought you might want to be better about it here.

Reply:
Well, I can delete here explicitly the deleted token, but, as in the other case we only want to delete the token if it is used (or like in this cases, if it is expired.